File: src/main/python/AoC2015_04.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 4
#

import sys
from hashlib import md5

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.common import spinner

Input = str
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return list(input_data)[0]

    def find_md5_starting_with_zeroes(self, seed: str, zeroes: int) -> int:
        i = 0
        val = seed
        target = "0" * zeroes
        while val[:zeroes] != target:
            i += 1
            spinner(i)
            str2hash = seed + str(i)
            val = md5(str2hash.encode()).hexdigest()  # nosec
        return i

    def part_1(self, seed: Input) -> Output1:
        return self.find_md5_starting_with_zeroes(seed, 5)

    def part_2(self, seed: Input) -> Output2:
        return self.find_md5_starting_with_zeroes(seed, 6)

    @aoc_samples(
        (
            ("part_1", "abcdef", 609043),
            ("part_1", "pqrstuv", 1048970),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2015, 4)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/cpp/2015/04/AoC2015_04.cpp
#include <assert.h>

#include <string>
#include <vector>

#include "../../aoc/aoc.hpp"
#include "../../aoc/md5/md5.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;

int findMd5StartingWithZeroes(const string& seed, const int zeroes) {
    for (int i = 0;; ++i) {
        const string& text = seed + to_string(i);
        MD5 md5;
        md5.update(text.c_str(), text.length());
        md5.finalize();
        const unsigned char* digest = md5.getDigest();
        int cnt = 0;
        for (int j = 0; j < zeroes / 2 + zeroes % 2; j++) {
            unsigned char ch = *(digest + j);
            if ((ch & 0xF0) == 0) {
                cnt++;
                if ((ch & 0x0F) == 0) {
                    cnt++;
                    continue;
                }
            }
            break;
        }
        if (cnt == zeroes) {
            return i;
        }
    }
}

int part1(const vector<string>& input) {
    return findMd5StartingWithZeroes(input[0], 5);
}

int part2(const vector<string>& input) {
    return findMd5StartingWithZeroes(input[0], 6);
}

const vector<string> TEST1 = {"abcdef"};
const vector<string> TEST2 = {"pqrstuv"};

void samples() {
    assert(part1(TEST1) == 609043);
    assert(part1(TEST2) == 1048970);
}

MAIN(2015, 4)
